src/malebolgia/paralgos.nim

```diff
@@ -0,0 +1,102 @@
+# (c) 2023 Andreas Rumpf
+
+import malebolgia
+
+template `@!`[T](data: openArray[T]; i: int): untyped =
+  cast[ptr UncheckedArray[T]](addr data[i])
+
+template parMap*[T](data: var openArray[T]; bulkSize: int; op: untyped) =
+  ##[ Applies `op` on every element of `data`, `op(data[i)` in parallel.
+  `bulkSize` specifies how many elements are processed sequentially
+  and not in parallel because sending a task to a different thread
+  is expensive. `bulkSize` should probably be larger than you think but
+  it depends on how expensive the operation `op` is.]##
```

Comment:
Maybe should change that to _start new thread_ instead of _send to thread_
Unless I got it wrong, then I would be glad if you could enlighten my misconceptions.

And [I'm trying to find this magic number](https://gist.github.com/hugosenari/dbc33d833593e3b21eb83299295580b1#notes).

Reply:
`spawn` DOES NOT start threads, it runs a task on the existing threadpool.

Reply:
> And [I'm trying to find this magic number](https://gist.github.com/hugosenari/dbc33d833593e3b21eb83299295580b1#notes).

But for which operation? Anyway, this is awesome! I am dreaming of a tool that tweaks bulkSize, numer of threads and thread queue sizes based on a workload. Maybe you can help?

Reply:
> a tool that tweaks bulkSize, numer of threads and thread queue sizes based on a workload.

Maybe the best we can do is provide a template like this:
```nim
benchParApply:
   # insert your workload
```
In the branch of my last PR I did [something like this](https://github.com/riinr/malebolgia/blob/spin-doctors/bench/parapply/tparapply_male.nim#L34), but is not polished, i.e. parameters are `define`.
-  bulkSize   {.intdefine.} = 2
-  operations {.intdefine.} = 3 * bulkSize
-  runs       {.intdefine.} = 1000
-  separeator {.strdefine.} = "\t"

Hint:
If you load the output of benchParApply with `pandas.read_table(sys.argv[1]).describe().transpose().to_csv()))`¹, you get output like this²:

```tsv
                         count  mean   std    min    25%   50%   75%    max
Setup                    1000   101    63     41     56    87    120    831
Serial                   1000   41     17     20     26    37    56     173
Spwn0000E0 - Epoch       1000   9857   10677  2334   4597  7322  12166  161291
Spwn0001E0 - Epoch       1000   10807  11054  2534   5502  8237  13308  173191
Spwn0002E0 - Epoch       1000   11466  11230  2686   6068  8860  13825  174548
Spwn0001E0 - Spwn0000E1  1000   908    1360   -3608  267   498   1140   22993
Spwn0002E0 - Spwn0001E1  1000   618    1255   -3788  163   359   798    15018
```

¹ [out2stat.py](https://gist.github.com/hugosenari/dbc33d833593e3b21eb83299295580b1#file-stats-py)
² long example [256ops_128spawns_8threads.csv](https://gist.github.com/hugosenari/dbc33d833593e3b21eb83299295580b1#file-256ops_128spawns_8threads-csv)